File: reference/rocketride-server/apps/vscode/src/agents/services.ts
// =============================================================================
// MIT License
// Copyright (c) 2026 Aparavi Software AG
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
// =============================================================================

/**
 * services.ts - Service Catalog Sync
 *
 * Syncs the server's service catalog to .rocketride/ in the workspace:
 *   1. Individual schema files:  .rocketride/schema/<component>.json
 *   2. Obsolete schema cleanup:  removes files not in the current catalog
 *   3. Master catalog:           .rocketride/services-catalog.json
 */

import * as vscode from 'vscode';
import { getLogger } from '../shared/util/output';
import { icons } from '../shared/util/icons';

/**
 * Write a file only if content has actually changed (line-ending normalized).
 */
async function writeIfChanged(uri: vscode.Uri, content: string): Promise<boolean> {
	try {
		const existing = Buffer.from(await vscode.workspace.fs.readFile(uri)).toString('utf8');
		if (existing.replace(/\r\n/g, '\n') === content.replace(/\r\n/g, '\n')) {
			return false;
		}
	} catch {
		// File doesn't exist — will create
	}
	await vscode.workspace.fs.writeFile(uri, Buffer.from(content, 'utf8'));
	return true;
}

/**
 * Extract the first sentence from an HTML description string.
 */
function firstSentence(description: string | undefined): string {
	if (!description) return '';
	// Strip HTML tags. Loop until idempotent so nested or malformed tags like
	// `<scri<x>pt>` collapse fully — a single regex pass would leave `<script>`
	// behind after consuming only the inner `<x>`. Defense-in-depth: the
	// output here lands in a JSON catalog file, not rendered as HTML, but
	// the consumer surface may grow.
	let stripped = description;
	let prev: string;
	do {
		prev = stripped;
		stripped = stripped.replace(/<[^>]*>/g, '');
	} while (stripped !== prev);
	const text = stripped.trim();
	// Take first sentence (ends with . ! or ?)
	const match = text.match(/^[^.!?]*[.!?]/);
	return match ? match[0].trim() : text;
}

/**
 * Sanitize a service name so it is safe to use as a filename.
 * Only `[a-zA-Z0-9._-]` characters are kept; everything else is replaced with `_`.
 * Each leading dot is replaced individually to preserve name uniqueness.
 */
function sanitizeServiceName(name: string): string {
	return name.replace(/[^a-zA-Z0-9._-]/g, '_').replace(/^\.+/, (match) => '_'.repeat(match.length));
}

/**
 * Return true only when `child` is strictly inside `parent` (same scheme + authority,
 * and child path starts with parent path followed by a `/`).
 * Used as a defense-in-depth guard after name sanitization.
 */
function isUnderDirectory(parent: vscode.Uri, child: vscode.Uri): boolean {
	if (child.scheme !== parent.scheme || child.authority !== parent.authority) {
		return false;
	}
	const parentPath = parent.path.endsWith('/') ? parent.path : parent.path + '/';
	return child.path.startsWith(parentPath);
}

/**
 * Sync service catalog data to .rocketride/ when the server sends services.
 *
 * 1. Write individual schema files: .rocketride/schema/<component>.json
 * 2. Remove obsolete schema files not in the current catalog
 * 3. Write master catalog: .rocketride/services-catalog.json
 */
export async function syncServiceCatalog(workspaceRoot: vscode.Uri, services: Record<string, unknown>): Promise<void> {
	const schemaDir = vscode.Uri.joinPath(workspaceRoot, '.rocketride', 'schema');
	await vscode.workspace.fs.createDirectory(schemaDir);

	const logger = getLogger();
	const serviceNames = Object.keys(services);

	// Step 1: Write individual schema files; track written filenames for cleanup.
	const expectedFiles = new Set<string>();
	for (const name of serviceNames) {
		const safeName = sanitizeServiceName(name);
		const schemaUri = vscode.Uri.joinPath(schemaDir, `${safeName}.json`);
		// Defense-in-depth: verify the resolved path is still within schemaDir.
		if (!isUnderDirectory(schemaDir, schemaUri)) {
			logger.output(`${icons.info} Skipped schema write for unsafe service name: ${name}`);
			continue;
		}
		expectedFiles.add(`${safeName}.json`);
		const json = JSON.stringify(services[name], null, 2);
		await writeIfChanged(schemaUri, json);
	}

	// Step 2: Remove obsolete schema files
	try {
		const entries = await vscode.workspace.fs.readDirectory(schemaDir);
		for (const [fileName, fileType] of entries) {
			if (fileType === vscode.FileType.File && !expectedFiles.has(fileName)) {
				await vscode.workspace.fs.delete(vscode.Uri.joinPath(schemaDir, fileName));
				logger.output(`${icons.info} Removed obsolete schema: ${fileName}`);
			}
		}
	} catch {
		// Directory listing failed — first run, nothing to clean
	}

	// Step 3: Build and write master catalog
	const catalog = serviceNames.map((name) => {
		const svc = services[name] as Record<string, unknown>;
		const entry: Record<string, unknown> = {
			name,
			classType: svc.classType ?? [],
			description: firstSentence(svc.description as string | undefined),
			lanes: svc.lanes ?? {},
		};
		if (svc.invoke !== undefined) {
			entry.invoke = svc.invoke;
		}
		return entry;
	});

	const catalogUri = vscode.Uri.joinPath(workspaceRoot, '.rocketride', 'services-catalog.json');
	await writeIfChanged(catalogUri, JSON.stringify(catalog, null, 2));
}

File: reference/rocketride-server/packages/client-python/src/rocketride/mixins/services.py
# MIT License
#
# Copyright (c) 2026 Aparavi Software AG
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

"""
Service Definitions for RocketRide Client.

This module provides methods to retrieve service (connector) definitions from
the RocketRide server. Service definitions include schemas, UI schemas, and
configuration metadata used for pipeline configuration and validation.

Usage:
    # Get all available service definitions
    services = await client.get_services()
    # services is a dict with 'services' and 'version' keys from the engine

    # Get a specific service by name
    ocr_schema = await client.get_service('ocr')
    if ocr_schema:
        print('OCR schema:', ocr_schema)
"""

from typing import Dict, Any, Optional

from ..core import DAPClient
from ..types.pipeline import PipelineConfig
from ..types.service import SERVICES_RESPONSE, SERVICE_DEFINITION, VALIDATION_RESULT


class ServicesMixin(DAPClient):
    """
    Provides service definition retrieval for the RocketRide client.

    This mixin adds get_services() and get_service() to fetch connector
    definitions from the server via the DAP rrext_services command.
    Definitions include schemas, UI schemas, and metadata.

    This is automatically included when you use RocketRideClient.
    """

    def __init__(self, **kwargs):
        """Initialize services functionality."""
        super().__init__(**kwargs)

    async def get_services(self) -> SERVICES_RESPONSE:
        """
        Retrieve all available service definitions from the server.

        Returns the full services structure from the engine, including a
        'services' dict (logical type -> definition) and 'version'.

        Returns:
            Dict containing 'services' (dict of service definitions) and
            'version'. Returns {} if the response has no body.

        Raises:
            RuntimeError: If the server returns an error.
        """
        return await self.call('rrext_services')

    async def get_service(self, service: str) -> Optional[SERVICE_DEFINITION]:
        """
        Retrieve a specific service definition by name.

        Args:
            service: Logical name of the service (e.g. 'ocr', 'embed', 'chat').

        Returns:
            The service definition dict, or None if not found.

        Raises:
            ValueError: If service is empty.
            RuntimeError: If the server returns an error (e.g. service not found).
        """
        if not service:
            raise ValueError('Service name is required')

        return await self.call('rrext_services', service=service)

    async def validate(
        self,
        pipeline: PipelineConfig,
        *,
        source: Optional[str] = None,
    ) -> VALIDATION_RESULT:
        """
        Validate a pipeline configuration.

        Sends the pipeline to the server for structural validation, checking
        component compatibility, connection integrity, and the resolved
        execution chain.

        Source resolution follows the same logic as :meth:`use`:
        1. Explicit ``source`` parameter (if provided)
        2. ``source`` field inside the pipeline config
        3. Implied source: the single component whose config.mode is 'Source'

        Args:
            pipeline: Pipeline configuration to validate.
            source: Optional override for the source component ID.

        Returns:
            Validation result containing errors and warnings.

        Raises:
            RuntimeError: If the server returns a validation error.

        Example:
            result = await client.validate(
                pipeline={'components': [...], 'project_id': '123'},
                source='webhook_1',
            )
        """
        try:
            kwargs: Dict[str, Any] = {'pipeline': pipeline}
            if source is not None:
                kwargs['source'] = source

            return await self.call('rrext_validate', **kwargs)
        except Exception as err:
            raise RuntimeError(f'Pipeline validation failed: {err}') from err
